be/src/storage/lake/meta_file.cpp

```diff
@@ -648,4 +658,132 @@ bool is_primary_key(const TabletMetadata& metadata) {
+        _pending_rowset_data.rowset_pb.CopyFrom(rowset_pb);
+    } else {
+        // Merge segments
+        for (int i = 0; i < rowset_pb.segments_size(); i++) {
+            _pending_rowset_data.rowset_pb.add_segments(rowset_pb.segments(i));
+        }
+        // Merge segment_size
+        for (int i = 0; i < rowset_pb.segment_size_size(); i++) {
+            _pending_rowset_data.rowset_pb.add_segment_size(rowset_pb.segment_size(i));
+        }
+        // Merge segment_encryption_metas
+        for (int i = 0; i < rowset_pb.segment_encryption_metas_size(); i++) {
+            _pending_rowset_data.rowset_pb.add_segment_encryption_metas(rowset_pb.segment_encryption_metas(i));
+        }
+        // Merge shared_segments
+        for (int i = 0; i < rowset_pb.shared_segments_size(); i++) {
+            _pending_rowset_data.rowset_pb.add_shared_segments(rowset_pb.shared_segments(i));
+        }
+    }
+
+    // Merge replace_segments
+    for (const auto& replace_seg : replace_segments) {
+        _pending_rowset_data.replace_segments[replace_seg.first] = replace_seg.second;
+    }
+
+    // Merge orphan_files
+    _pending_rowset_data.orphan_files.insert(_pending_rowset_data.orphan_files.end(), orphan_files.begin(),
+                                             orphan_files.end());
+
+    // Merge delete files
+    _pending_rowset_data.dels.insert(_pending_rowset_data.dels.end(), dels.begin(), dels.end());
+    _pending_rowset_data.del_encryption_metas.insert(_pending_rowset_data.del_encryption_metas.end(),
+                                                     del_encryption_metas.begin(), del_encryption_metas.end());
+}
+
+void MetaFileBuilder::set_final_rowset() {
+    if (_pending_rowset_data.rowset_pb.segments_size() == 0 && _pending_rowset_data.dels.empty()) {
+        return; // Nothing to do
+    }
+
+    auto rowset = _tablet_meta->add_rowsets();
+    rowset->CopyFrom(_pending_rowset_data.rowset_pb);
+
+    auto segment_size_size = rowset->segment_size_size();
+    auto segment_file_size = rowset->segments_size();
+    LOG_IF(ERROR, segment_size_size > 0 && segment_size_size != segment_file_size)
+            << "segment_size size != segment file size, tablet: " << _tablet.id() << ", rowset: " << rowset->id()
+            << ", segment file size: " << segment_file_size << ", segment_size size: " << segment_size_size;
+
+    // Apply replace_segments
+    for (const auto& replace_seg : _pending_rowset_data.replace_segments) {
+        rowset->set_segments(replace_seg.first, replace_seg.second.path);
+        rowset->set_segment_size(replace_seg.first, replace_seg.second.size.value());
+        if (replace_seg.first < rowset->segment_encryption_metas_size()) {
+            rowset->set_segment_encryption_metas(replace_seg.first, replace_seg.second.encryption_meta);
+        }
+        rowset->clear_bundle_file_offsets(); // clear shared file offsets, since we rewrite segments.
+    }
+
+    rowset->set_id(_tablet_meta->next_rowset_id());
```

Comment:
`next rowset id` has already been updated in `log applier`.
Is there a problem with the rowset using this value?